toolchain/check/subst.cpp

```diff
@@ -0,0 +1,241 @@
+  PushOperand(context, worklist, kinds.second, inst.arg1());
+}
+
+// Pops the specified operand from the worklist and returns it.
+static auto PopOperand(Context& context, Worklist& worklist, SemIR::IdKind kind,
+                       int32_t arg) -> int32_t {
+  switch (kind) {
+    case SemIR::IdKind::For<SemIR::InstId>:
+      return worklist.Pop().index;
+    case SemIR::IdKind::For<SemIR::TypeId>:
+      return context.GetTypeIdForTypeInst(worklist.Pop()).index;
+    case SemIR::IdKind::For<SemIR::InstBlockId>: {
+      auto old_inst_block_id = static_cast<SemIR::InstBlockId>(arg);
+      auto size = context.inst_blocks().Get(old_inst_block_id).size();
+      SemIR::CopyOnWriteInstBlock new_inst_block(context.sem_ir(),
+                                                 old_inst_block_id);
+      for (auto i : llvm::index_range(0, size)) {
+        new_inst_block.Set(size - i - 1, worklist.Pop());
+      }
+      return new_inst_block.id().index;
+    }
+    case SemIR::IdKind::For<SemIR::TypeBlockId>: {
+      auto old_type_block_id = static_cast<SemIR::TypeBlockId>(arg);
+      auto size = context.type_blocks().Get(old_type_block_id).size();
+      SemIR::CopyOnWriteTypeBlock new_type_block(context.sem_ir(),
+                                                 old_type_block_id);
+      for (auto i : llvm::index_range(0, size)) {
+        new_type_block.Set(size - i - 1,
+                           context.GetTypeIdForTypeInst(worklist.Pop()));
+      }
+      return new_type_block.id().index;
+    }
+    default:
+      return arg;
+  }
+}
+
+// Pops the operands of the specified instruction off the worklist and rebuilds
+// the instruction with the updated operands.
+static auto Rebuild(Context& context, Worklist& worklist, SemIR::InstId inst_id)
+    -> SemIR::InstId {
+  auto inst = context.insts().Get(inst_id);
+  auto kinds = inst.ArgKinds();
+
+  // Note that we pop in reverse order because we pushed them in forwards order.
+  int32_t arg1 = PopOperand(context, worklist, kinds.second, inst.arg1());
+  int32_t arg0 = PopOperand(context, worklist, kinds.first, inst.arg0());
+  if (arg0 == inst.arg0() && arg1 == inst.arg1()) {
+    return inst_id;
+  }
+
+  inst.SetArgs(arg0, arg1);
+  auto result_id = TryEvalInst(context, SemIR::InstId::Invalid, inst);
+  CARBON_CHECK(result_id.is_constant())
+      << "Substitution into constant produced non-constant";
+  return result_id.inst_id();
+}
+
+auto SubstConstant(Context& context, SemIR::ConstantId const_id,
+                   Substitutions substitutions) -> SemIR::ConstantId {
+  CARBON_CHECK(const_id.is_constant()) << "Substituting into non-constant";
+
+  if (substitutions.empty()) {
+    // Nothing to substitute.
+    return const_id;
+  }
+
+  if (!const_id.is_symbolic()) {
+    // A template constant can't contain a reference to a symbolic binding.
+    return const_id;
+  }
+
+  Worklist worklist;
+  worklist.Push(const_id.inst_id());
+  worklist.Get(0).next_index = -1;
```

Comment:
Maybe move this to a constructor which can do it without a separate `Get`?

Reply:
Done.